torch/csrc/jit/script/python_sugared_value.cpp

```diff
@@ -32,71 +32,76 @@ FunctionSchema PythonValue::getSchema(
     const size_t n_binders,
     const SourceRange& loc) {
   auto annotations = py::module::import("torch.jit.annotations");
-  const auto fn_to_get_signature =
-      moduleSelf_ ? py::getattr(self, "original_fn") : self;
-  auto signature = annotations.attr("get_signature")(
-      fn_to_get_signature, rcb ? *rcb : py::none(), loc);
+  const auto callable = moduleSelf_ ? py::getattr(self, "original_fn") : self;
+
+  // Make sure the function is not a class instantiation
+  annotations.attr("check_fn")(callable, loc);
+  auto is_vararg = py::cast<bool>(annotations.attr("is_vararg")(callable));
+
+  auto signature =
+      annotations.attr("get_signature")(callable, rcb ? *rcb : py::none(), loc);
   std::vector<Argument> args, rets;
 
+  auto py_param_names = annotations.attr("get_param_names")(callable, n_args);
+  auto param_names = py::cast<std::vector<std::string>>(py_param_names);
+  auto names_it = param_names.begin();
   if (moduleSelf_) {
-    args.push_back(Argument("self", moduleSelf_->type(), {}, {}, false));
-  }
-  // We may mutate this if we can determine the number of args from Python
-  // introspection.
-  size_t actual_n_args = moduleSelf_ ? n_args + 1 : n_args;
-  if (!signature.is_none()) {
-    std::vector<TypePtr> arg_types;
-    TypePtr ret_type;
-    std::tie(arg_types, ret_type) =
-        py::cast<std::pair<std::vector<TypePtr>, TypePtr>>(signature);
-    args.reserve(arg_types.size());
-    size_t idx = 0; // Fake argument names by putting in the index
-    for (auto& arg_type : arg_types) {
-      args.push_back(
-          Argument(std::to_string(idx++), std::move(arg_type), {}, {}, false));
-    }
-    rets.push_back(Argument("0", std::move(ret_type), {}, {}, false));
-  } else {
-    // Create a default signature using what information we have
-
-    // First see if we can introspect the number of function parameters
-    // irrespective of the presence of explicit type annotations
-    auto num_params =
-        annotations.attr("get_num_params")(fn_to_get_signature, loc);
-    if (!num_params.is_none()) {
-      // Return a signature with the correct number of params according to the
-      // Python function. The error handling in call() will catch any mismatch
-      // later.
-      actual_n_args = py::cast<size_t>(num_params);
-      if (moduleSelf_) {
-        TORCH_INTERNAL_ASSERT(actual_n_args > 0);
-        --actual_n_args;
-      }
-    }
-    // Construct the default signature: all arguments and returns will be
-    // DynamicType
-    args.reserve(actual_n_args);
-    for (size_t i = 0; i < actual_n_args; ++i) {
-      args.push_back(
-          Argument(std::to_string(i), TensorType::get(), {}, {}, false));
+    // If there is a `self` parameter on the callable, skip it on the names list
+    ++names_it;
```

Comment:
The `self` logic is fine for probably all cases. It would be safer to check if it were a method but probably not needed.